lib/puppetserver/ca/sign_action.rb

```diff
@@ -0,0 +1,175 @@
+          if requested_certnames.nil?
+            return 1
+          end
+        else
+          requested_certnames = input['certname']
+        end
+
+        success = sign_requested_certs(requested_certnames, puppet.settings)
+        return success ? 0 : 1
+      end
+
+      def get_certificate_statuses(settings)
+        url = HttpUtilities.make_ca_url(settings[:ca_server], settings[:ca_port],
+                                        'certificate_statuses', 'any_key')
+        HttpUtilities.with_connection(url, settings) do |connection|
+          connection.get(url)
+        end
+      end
+
+      def sign_cert(certname,settings)
+        url = HttpUtilities.make_ca_url(settings[:ca_server], settings[:ca_port],
+                                        'certificate_status')
+        HttpUtilities.with_connection(url, settings) do |connection|
+          url.resource_name = certname
+          connection.put(BODY, url)
+        end
+      end
+
+      def get_all_certs(settings)
+        result = get_certificate_statuses(settings)
+
+        unless result.is_a? Net::HTTPOK
+            @logger.err 'Error:'
+            @logger.err "#{result.inspect}"
+            return nil
+        end
+        return result
+      end
+
+      def select_pending_certs(get_result)
+        requested_certnames = JSON.parse(get_result).select{|e| e["state"] == "requested"}.map{|e| e["name"]}
+
+        if requested_certnames.empty?
+          @logger.err 'Error'
+          @logger.err "No waiting certificate requests to sign"
+          return nil
+        end
+
+        return requested_certnames
+      end
+
+      def get_all_pending_certs(settings)
+        result = get_all_certs(settings)
+        if result
+          select_pending_certs(result.body)
+        end
+      end
+
+      def sign_requested_certs(certnames,settings)
+        success = true
+        certnames.each do |certname|
+          result = sign_cert(certname,settings)
+          case result.code
+          when '204'
+            @logger.inform "Signed certificate for #{certname}"
+          when '404'
+            @logger.err 'Error:'
+            @logger.err "Could not find certificate for #{certname}"
+            success = false
+          else
+            @logger.err 'Error:'
+            @logger.err "#{result.inspect}"
+            success = false
+          end
+        end
+        return success
+      end
+
+      def check_flag_usage(results)
+        if results['certname'] && results['all']
+          '--all and --certname cannot be used together'
+        elsif !results['certname'] && !results['all']
+          'No arguments given'
+        elsif results['certname'] && results['certname'].include?('--all')
+          'Cannot use --all with --certname. If you actually have a certificate request ' +
+                          'for a certifcate named --all, you need to use the HTTP API.'
+        end
```

Comment:
I lol'd. Are dashes valid leading characters for a certname?

Reply:
I'm not sure about that, actually. If not, we could lose some of the snark in this error string